go/vt/mysqlctl/backup.go

```diff
@@ -218,34 +218,25 @@ func removeExistingFiles(cnf *Mycnf) error {
 	return nil
 }
 
+// ShouldRestore checks whether a database with tables already exists
+// and returns whether a restore action should be performed
+func ShouldRestore(ctx context.Context, params RestoreParams) (bool, error) {
+	if params.DeleteBeforeRestore || RestoreWasInterrupted(params.Cnf) {
+		return true, nil
+	}
+	params.Logger.Infof("Restore: No %v file found, checking no existing data is present", RestoreState)
+	// Wait for mysqld to be ready, in case it was launched in parallel with us.
+	// If this doesn't succeed, assume we should attempt a restore
+	if err := params.Mysqld.Wait(ctx, params.Cnf); err != nil {
+		return true, err
+	}
+	return checkNoDB(ctx, params.Mysqld, params.DbName)
+}
+
 // Restore is the main entry point for backup restore.  If there is no
 // appropriate backup on the BackupStorage, Restore logs an error
 // and returns ErrNoBackup. Any other error is returned.
 func Restore(ctx context.Context, params RestoreParams) (*BackupManifest, error) {
-
-	if !params.DeleteBeforeRestore {
-		params.Logger.Infof("Restore: Checking if a restore is in progress")
-		if !RestoreWasInterrupted(params.Cnf) {
-			params.Logger.Infof("Restore: No %v file found, checking no existing data is present", RestoreState)
-			// Wait for mysqld to be ready, in case it was launched in parallel with us.
-			if err := params.Mysqld.Wait(ctx, params.Cnf); err != nil {
-				return nil, err
-			}
-
-			ok, err := checkNoDB(ctx, params.Mysqld, params.DbName)
-			if err != nil {
-				return nil, err
-			}
-			if !ok {
-				params.Logger.Infof("Auto-restore is enabled, but mysqld already contains data. Assuming vttablet was just restarted.")
-				if err = PopulateMetadataTables(params.Mysqld, params.LocalMetadata, params.DbName); err == nil {
-					err = ErrExistingDB
```

Comment:
It seems like we no longer return `ErrExistingDB` anywhere, which makes sense; `ShouldRestore()` returning `false` is a cleaner way to indicate that condition. We should be able to remove all other references to `ErrExistingDB` as well, right?

Reply:
I looked at all the refs to `ErrExistingDB`. It is true that none of the code paths that act on that error value will be hit any more (because either we return earlier on error, or because Restore no longer returns that error code). However, we should keep this specific error so that the exit from a vttablet startup produces the same error message as before.

Reply:
> none of the code paths that act on that error value will be hit any more

In addition, there are no longer any code paths that return this error. I think this was the only place, and it isn't needed here anymore.

> we should keep this specific error so that the exit from a vttablet startup produces the same error message as before

I don't think this error ever leaked out beyond whatever function directly called `Restore()`. It was not really treated as an error, so it was always swallowed immediately. In the new world, we correctly represent this as a non-error state, when `ShouldRestore()` returns `false, nil`.

Reply:
I thought that RestoreData should return ErrExistingDB. PTAL at the latest commit.